eo-maven-plugin/src/main/java/org/eolang/maven/ProbeMojo.java

```diff
@@ -0,0 +1,246 @@
+ * catalog.
+ *
+ * @since 0.28.11
+ * @checkstyle CyclomaticComplexityCheck (300 lines)
+ */
+@Mojo(
+    name = "probe",
+    defaultPhase = LifecyclePhase.PROCESS_SOURCES,
+    threadSafe = true
+)
+public final class ProbeMojo extends SafeMojo {
+
+    /**
+     * The Git hash to pull objects from, in objectionary.
+     *
+     * @since 0.21.0
+     */
+    @SuppressWarnings("PMD.ImmutableField")
+    @Parameter(property = "eo.tag", required = true, defaultValue = "master")
+    private String tag = "master";
+
+    /**
+     * Target directory.
+     *
+     * @checkstyle MemberNameCheck (7 lines)
+     */
+    @Parameter(property = "eo.home")
+    @SuppressWarnings("PMD.ImmutableField")
+    private Path outputPath = Paths.get(System.getProperty("user.home")).resolve(".eo");
+
+    /**
+     * Read hashes from local file.
+     *
+     * @checkstyle MemberNameCheck (7 lines)
+     */
+    @Parameter(property = "offlineHashFile")
+    private Path offlineHashFile;
+
+    /**
+     * Return hash by pattern.
+     * -DofflineHash=0.*.*:abc2sd3
+     * -DofflineHash=0.2.7:abc2sd3,0.2.8:s4se2fe
+     *
+     * @checkstyle MemberNameCheck (7 lines)
+     */
+    @Parameter(property = "offlineHash")
+    private String offlineHash;
+
+    /**
+     * The objectionary.
+     */
+    @SuppressWarnings("PMD.ImmutableField")
+    private Objectionary objectionary;
+
+    @Override
+    public void exec() throws IOException {
+        if (!new Online().value()) {
+            Logger.warn(
+                this, "We are offline: skipping probes"
+            );
+            return;
+        }
+        final Collection<Tojo> tojos = this.scopedTojos().select(
+            row -> row.exists(AssembleMojo.ATTR_XMIR2)
+                && !row.exists(AssembleMojo.ATTR_PROBED)
+        );
+        final CommitHash hash = new ChResolve(
+            this.offlineHashFile, this.offlineHash, this.tag
+        );
+        if (this.objectionary == null) {
+            this.objectionary = new OyFallbackSwap(
+                new OyHome(
+                    new ChNarrow(hash),
+                    this.outputPath
+                ),
+                new OyRemote(hash),
+                this.forceUpdate()
+            );
+        }
+        final Collection<String> probed = new HashSet<>(1);
+        for (final Tojo tojo : tojos) {
+            final Path src = Paths.get(tojo.get(AssembleMojo.ATTR_XMIR2));
+            final Collection<String> names = this.probe(src);
+            int count = 0;
+            for (final String name : names) {
+                if (!this.objectionary.contains(name)) {
+                    continue;
+                }
+                ++count;
+                final Tojo ftojo = this.scopedTojos().add(name);
+                if (!ftojo.exists(AssembleMojo.ATTR_VERSION)) {
+                    ftojo.set(AssembleMojo.ATTR_VERSION, "*.*.*");
+                }
+                ftojo.set(AssembleMojo.ATTR_PROBED_AT, src);
```

Comment:
@MikhailLipanin I believe in this case `ATTR_DISCOVERED_AT` can be reused. It has the same meaning as `ATTR_PROBED_AT`: the origin source where the object was found. I don't see why would we need to distinguish between discovery by discover step or by probe step.

Reply:
done